### PR TITLE
Create entities lists as recommendation feature

### PR DESCRIPTION
**Pull request by jum-s**
solve #184

in this PR:
- lists are the new shelves
- selections are the new items

main differences with existing shelves/items:
- no visibility on selections, which avoids to filter list selections (shelves have to filter private items)
- without visibility, selections can only be created, deleted and displayed through `lists` endpoint
- getting a list with selections (aka `byIdsWithSelections`) returns the whole document selection (shelves return items ids)

opinions:

- Selection takes `list` and not several `lists` in order to be able to extend the selection model with further data keys (details) but stay specific to a list. Usecase is unknown yet (who is going to be comfortable with same selection data accross several lists), but merging existing different lists selections with a same uri may be easier than spliting them.

### CODE REVIEW

File: server/controllers/lists/remove_selections.js
Comment (maxlath) [inline server/controllers/lists/remove_selections.js:22]:
```suggestion
  if (selectionsToDelete.length === 0) {
```

File: server/controllers/lists/update.js
Comment (maxlath) [inline server/controllers/lists/update.js:4]: The other controllers use the `id` attribute name, which makes that one inconsistent, no? That inconsistency seems to have been imported from shelves, maybe it could be made consistent there too, but that would be a breaking change that would require a client update, so maybe in another branch to not complexify this PR's release process

File: server/models/list.js
Comment (maxlath) [inline server/models/list.js:8]: those model functions could have some unit tests

File: tests/api/fixtures/lists.js
Comment (maxlath) [inline tests/api/fixtures/lists.js:28]: why not simply
```suggestion
    const { list } = await customAuthReq(user, 'post', `${endpoint}create`, listData)
    return { list, user }
  },
```
?

File: tests/api/lists/add_selections.test.js
Comment (maxlath) [inline tests/api/lists/add_selections.test.js:62]: a test do set the behavior of trying to add a list element with a uri that is already in the list would be welcome

**Review by maxlath — CHANGES_REQUESTED** [review]
Exciting! We are finally getting there :sweat_smile: 

Some general remarks that come to mind as I'm approaching this PR:
- what's the rational for the `selection` naming? We might have discussed that earlier but I can't recover where or recall the argument. I liked `element` as a more neutral option, but could be convinced otherwise
- things might get clearer when we get to client implementation, but as this is not breaking any existing feature, I would be fine with merging it once this is approved, without waiting for the client to be implemented, in order to limit the possible conflicts with the incoming items transition from `listing` to `visibility`

File: server/lib/filter_visible_docs.js
Comment (jum-s) [reply server/lib/filter_visible_docs.js:25]: i remember having a conversation that stated that `owner` means owning something in the physical world (aka items) and user being too general, so i find creator quite explicit and future-proof (this creator is already a refactor, it originally started with user...)

>  Is that creator attribute in anticipation of a possible future where the list owner isn't just the creator?

yes

**Comment by jum-s** [comment]
> what's the rational for the selection naming?

To me, naming is to go to the more acceptable specifics.

First, why List ? Its named as a second choice because, to me, what users (humans) are doing with a List is exactly what editors (also humans, but considered as Entities in Inventaire) are doing with a Collection. Since Collection is reserved for Entities already, List it is. But a Collection would have fit better to me: you make a distinction, a separation between books (a curation one may also say) between the mass of existing books and the ones of this special group/list/collection (so its more than a mere list, as its neither a Todo nor an inventory, for ex).
So, if you agree that Collection is more specific (so better) than List. Then, with the same logic, a Selection is more specific than Element, as this name says that there is a distinction/separation made by a user. The user have selected actively some entities, aka a Selection.

File: server/controllers/lists/add_selections.js
Comment (jum-s) [reply server/controllers/lists/add_selections.js:6]: > as entities are the only documents to have uris, 

yeah thats what i was thinking, lists and selections are so heavily related to entities that i thought its acceptable to accept uris. 

The consistency was that the sanitization parameters do not offer `entities` as a possibility, so i went for the already existing `uris`.

I would say its fine like that

File: server/controllers/lists/by_creators.js
Comment (jum-s) [reply server/controllers/lists/by_creators.js:15]: fixed up in "add pagination to endpoints by-creators and by-entities"

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js:15]: Yep, decision is to be specific of Selection a dedicated to a List. It will be more future proofed (and easier to move back to mutualize selection in several list).
This has the advantage to be comment specific later: if i add a Les Misérables comment which belongs to a "Kids List", i dont want this comment to be visible in the Adult List which also having Les Misérables.

Comment (jum-s) [reply server/controllers/lists/by_entities.js:15]: fixup! add Selection as sub-List element

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js:31]: get an array for each requested uris

fixup! add Selection as sub-List element

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js:32]: fixup! add Selection as sub-List element

File: tests/api/lists/by_entities.js
Comment (jum-s) [reply tests/api/lists/by_entities.js:1]: done in "fix tests filename"

File: server/controllers/lists/by_id.js
Comment (jum-s) [reply server/controllers/lists/by_id.js:1]: it has some minor advantages of not passing the param `withSelection`, to not warn on `checkNotFoundList`. But the main purpose is to allow Selection pagination, otherwise `by-ids` needs to be paginated...

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js:18]: > confusing renaming

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js:34]: yeah, this is a remaining of when you could also have a Lists specific selections by entities

fixup! add Selection as sub-List element

File: server/controllers/lists/create.js
Comment (jum-s) [reply server/controllers/lists/create.js:6]: fixed in "prevent sharing default visibility array between objects"

File: server/controllers/lists/lib/lists.js
Comment (jum-s) [reply server/controllers/lists/lib/lists.js:71]: i dont consider async/await to be always preferable to old fashoned promise. For example, your suggestion adds more chars (and more parenthesis), which makes it less readable to me.
 tap has the advantage of directly knowing before reading the line that the next line will use the previous line result, which is quite nice in my opinion

File: tests/api/lists/add_selections.test.js
Comment (jum-s) [reply tests/api/lists/add_selections.test.js:13]: fixup! add Selection as sub-List element

File: server/controllers/lists/remove_selections.js
Comment (jum-s) [reply server/controllers/lists/remove_selections.js:10]: i imagined the client only sending uris and server knows what to do (but since client is not coded yet, it might be too far fetched)
I dont think its less future proof, since, in the case you mention, extending API to accept a selectionId params should be enough for clients.
maybe we could merge like this now, and add an update after client implementation?

File: server/controllers/items/lib/queries_commons.js
Comment (jum-s) [reply server/controllers/items/lib/queries_commons.js:10]: removed filter in fixup! add pagination to endpoints by-creators and by-entities

File: server/controllers/lists/delete_by_ids.js
Comment (jum-s) [reply server/controllers/lists/delete_by_ids.js:18]: done in fixup! add Selection as sub-List element

File: server/controllers/lists/update.js
Comment (jum-s) [reply server/controllers/lists/update.js:4]: joys of copy/pasting

!fixup implement basic CRUD

File: tests/api/fixtures/lists.js
Comment (jum-s) [reply tests/api/fixtures/lists.js:28]: !fixup implement basic CRUD

File: server/controllers/lists/remove_selections.js
Comment (jum-s) [reply server/controllers/lists/remove_selections.js:21]: done in 43c0c887

File: server/controllers/lists/lib/lists.js
Comment (jum-s) [reply server/controllers/lists/lib/lists.js:51]: fixed by 43c0c887

File: server/controllers/lists/lib/lists.js
Comment (jum-s) [reply server/controllers/lists/lib/lists.js:54]: done in 43c0c887

File: server/controllers/lists/lib/selections.js
Comment (jum-s) [reply server/controllers/lists/lib/selections.js]: fixup! add Selection as sub-List element

File: server/controllers/lists/lib/selections.js
Comment (jum-s) [reply server/controllers/lists/lib/selections.js]: fixup! add Selection as sub-List element

File: server/controllers/lists/remove_selections.js
Comment (jum-s) [reply server/controllers/lists/remove_selections.js]: fixup! add Selection as sub-List element

to not block the PR

File: tests/api/lists/add_selections.test.js
Comment (jum-s) [reply tests/api/lists/add_selections.test.js:61]: done in 43c0c887

File: server/models/list.js
Comment (jum-s) [reply server/models/list.js:8]: i dont really get the value of thos model tests for such a straight forward endpoint

but here you go 7c31b287

File: server/controllers/lists/by_entities.js
Comment (jum-s) [reply server/controllers/lists/by_entities.js]: has this been fixed with 0c58c80d ?

File: server/controllers/lists/add_selections.js
Comment (maxlath) [reply server/controllers/lists/add_selections.js:6]: the `uris` parameter is actually already used by `/api/items?action=by-entities`, so there is at least a precedent

File: server/controllers/lists/by_entities.js
Comment (maxlath) [reply server/controllers/lists/by_entities.js]: I can't recall what that "confusing renaming"  was referring to, so I guess yes :sweat_smile:

**Comment by maxlath** [comment]
My main issue with the "selection" naming is that it seems prone to confusion: when one talks of a list selection, it needs disambiguation as it might be understood as a selection of lists. The denomination "element" does not have that problem: a list element is very clearly a sub-part of that list. Furthermore, the term selection can generally be used to refer to several elements being selected:
- "here is my music selection" => a playlist
- "the national selection" => a team

The list goes on, cf https://en.wiktionary.org/wiki/selection: a selection can be either one or several things. That Wiktionary article also brought my attention to an even more problematic case: [Selection (user interface)](https://en.wikipedia.org/wiki/Selection_(user_interface)) *a selection is a list of items on which [user operations](https://en.wikipedia.org/wiki/Command_(computing)) will take place*.

The term "element"  does not have that ambiguity https://en.wiktionary.org/wiki/element (at one exception, that rare definition n°10. ;))

File: server/controllers/lists/lib/lists.js
Comment (maxlath) [reply server/controllers/lists/lib/lists.js:73]: I would agree that there are cases where old-fashioned promise chains are more readable, but here I would say that's not the case: `.then(db.bulkDelete)` leaves an ambiguity on whether returning `db.bulkDelete` return value is part of the function's contract. Using `await db.bulkDelete(lists)` removes the ambiguity, has it explicitly does not return anything.

**Comment by jum-s** [comment]
Your critics of selection are understandable, although i would like to keep rooted the code, with new terms selected in relation with other terms current usage in the codebase. General meaning is less important than the capacity for devs to clearly disambiguate when reading code.

Im opened to renaming Selection -> Element, since "element" is only used once on the server code. But List is used dozens of times (for the server part only), and is leading to longer naming disambiguation. We will be puzzled each time we need to name a function that manipulates arrays. I would suggest that the UI would still display the term "list", but code would gain clarity and improve developer happiness if other terms refers to those list. Anything but list : you expressed resistance to past suggestions (`collage`, 'user collection', 'bibliography') here are new (but less fitting) suggestions : 'publication', 'user list', 'folder', 'catalogue'

**Pull request merged**
